akka-stream/src/main/scala/akka/stream/ssl/SslTlsCipher.scala

```diff
@@ -1,376 +1,59 @@
-    }
-    val result = doWrap(temporaryBuffer)
-    val rs = result.getStatus
-    if (tracing) log.debug("## Encrypting {}", rs)
-    val hs = result.getHandshakeStatus
-    val next = rs match {
-      case OK ⇒
-        if (hs == NOT_HANDSHAKING) {
-          writeCipherTextOutboundBytes()
-          engineRunningPhase
-        } else handshakePhase(hs)
-      case CLOSED           ⇒ if (!engine.isInboundDone) decryptionPhase else completedPhase
-      case BUFFER_OVERFLOW  ⇒ throw new IllegalStateException // the SslBufferPool should make sure that buffers are never too small
-      case BUFFER_UNDERFLOW ⇒ throw new IllegalStateException // should never appear as a result of a wrap
-    }
-    nextPhase(next)
-  }
-
-  val decryptionPhase: TransferPhase = TransferPhase(canDecrypt) { () ⇒
-    if (tracing) log.debug("### Decrypting")
-    if (inboundCipherTextInput.NeedsInput.isReady) {
-      val elem = inboundCipherTextInput.dequeueInputElement().asInstanceOf[ByteString]
-      enqueueCipherInputBytes(elem)
-    }
-    val result = doUnwrap(temporaryBuffer)
-    val rs = result.getStatus
-    if (tracing) log.debug("## Decrypting {}", rs)
-    val hs = result.getHandshakeStatus
-    val next = rs match {
-      case OK ⇒
-        if (hs == NOT_HANDSHAKING) {
-          writePlaintextInboundBytes()
-          engineRunningPhase
-        } else handshakePhase(hs)
-      case CLOSED           ⇒ if (!engine.isOutboundDone) encryptionPhase else completedPhase
-      case BUFFER_OVERFLOW  ⇒ throw new IllegalStateException // the SslBufferPool should make sure that buffers are never too small
-      case BUFFER_UNDERFLOW ⇒ throw new IllegalStateException // should never appear as a result of a wrap
-    }
-    nextPhase(next)
-  }
-
-  nextPhase(waitForHandshakeStartPhase)
-
-  override def preStart() {
-    val plainTextInput = inboundSessionOutput.asInstanceOf[Publisher[SslTlsCipher.InboundSession]]
-    val plainTextOutput = new SubstreamSubscriber[ByteString](self, outboundPlaintextInput.key)
-    val cipherTextInput = new SubstreamSubscriber[ByteString](self, inboundCipherTextInput.key)
-    val cipherTextOutput = outboundCipherTextOutput.asInstanceOf[Publisher[ByteString]]
-    requester ! SslTlsCipher(plainTextInput, plainTextOutput, cipherTextInput, cipherTextOutput)
-  }
-
-  override def receive = inputSubstreamManagement orElse outputSubstreamManagement
+  def withParameters(p: SSLParameters) = default.copy(sslParameters = Some(p))
+}
 
-  protected def fail(e: Throwable): Unit = {
-    // FIXME: escalate to supervisor
-    if (tracing) log.debug("fail {} due to: {}", self, e.getMessage)
-    failInputs(e)
-    failOutputs(e)
-    context.stop(self)
-  }
+case class SendBytes(bytes: ByteString) extends SslTlsOutbound
 
-  override protected def pumpFailed(e: Throwable): Unit = fail(e)
+sealed abstract class ClientAuth
+object ClientAuth {
+  case object None extends ClientAuth
+  case object Want extends ClientAuth
```

Comment:
Hmm, I don’t think we should start documenting the JSSE APIs—these are explained in the JDK JavaDocs. But it is probably alright to put in a pointer.